File: python/mpca/naive_pca.py
"""Naive PC1 aggregation on the observed correlation matrix."""

import numpy as np


def naive_pca(S_hat: np.ndarray) -> dict:
    """
    Standardize sub-index scores and extract the leading PC.

    Parameters
    ----------
    S_hat : np.ndarray, shape (N, K)
        Imputed sub-index score matrix.

    Returns
    -------
    dict with keys:

    ``col_means`` : np.ndarray, shape (K,)
        Column means of *S_hat*.
    ``col_sds`` : np.ndarray, shape (K,)
        Column standard deviations of *S_hat*.
    ``S_tilde`` : np.ndarray, shape (N, K)
        Standardized score matrix.
    ``R_obs`` : np.ndarray, shape (K, K)
        Sample correlation matrix :math:`(N-1)^{-1}\\tilde{S}'\\tilde{S}`.
    ``lambda1`` : float
        Leading eigenvalue.
    ``ell_hat`` : np.ndarray, shape (K,)
        PC1 loadings (all positive by sign convention).
    ``w_hat`` : np.ndarray, shape (K,)
        Regression scoring coefficients
        :math:`\\hat{\\mathbf{w}} = \\hat{\\mathbf{R}}_{\\text{obs}}^{-1}\\hat{\\boldsymbol{\\ell}}`.
    ``f_hat`` : np.ndarray, shape (N,)
        Naive composite scores :math:`\\tilde{S}\\hat{\\mathbf{w}}`.
    ``var_explained`` : float
        Proportion of variance explained by PC1.
    """
    N, K = S_hat.shape

    col_means = S_hat.mean(axis=0)
    col_sds = S_hat.std(axis=0, ddof=1)

    S_tilde = (S_hat - col_means) / col_sds

    # Sample correlation matrix
    R_obs = (S_tilde.T @ S_tilde) / (N - 1)

    # Leading eigenpair (eigh returns ascending order)
    eigenvalues, eigenvectors = np.linalg.eigh(R_obs)
    idx = np.argsort(eigenvalues)[::-1]
    eigenvalues = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]

    lambda1 = eigenvalues[0]
    ell_hat = eigenvectors[:, 0]

    # Sign convention: sum of loadings positive
    if ell_hat.sum() < 0:
        ell_hat = -ell_hat

    # Regression scoring coefficients
    w_hat = np.linalg.solve(R_obs, ell_hat)

    # Composite scores
    f_hat = S_tilde @ w_hat

    var_explained = lambda1 / eigenvalues.sum()

    return {
        "col_means": col_means,
        "col_sds": col_sds,
        "S_tilde": S_tilde,
        "R_obs": R_obs,
        "lambda1": lambda1,
        "ell_hat": ell_hat,
        "w_hat": w_hat,
        "f_hat": f_hat,
        "var_explained": var_explained,
    }

File: r/R/naive_pca.R
#' Naive PC1 Aggregation
#'
#' Standardizes the N×K score matrix, computes the sample correlation matrix,
#' extracts the leading eigenpair (PC1), derives regression scoring
#' coefficients, and scores each observation on the naive composite.
#'
#' @param S_hat N×K numeric matrix of imputed sub-index scores.
#'
#' @return A named list:
#'   \itemize{
#'     \item `col_means`: length-K vector of column means.
#'     \item `col_sds`: length-K vector of column standard deviations.
#'     \item `S_tilde`: N×K standardized score matrix.
#'     \item `R_obs`: K×K sample correlation matrix.
#'     \item `lambda1`: leading eigenvalue.
#'     \item `ell_hat`: length-K PC1 loading vector (all positive by
#'       convention).
#'     \item `w_hat`: length-K regression scoring coefficients
#'       (\eqn{\hat{\mathbf{w}} = \hat{\mathbf{R}}_\text{obs}^{-1}\hat{\boldsymbol{\ell}}}).
#'     \item `f_hat`: length-N naive composite scores.
#'     \item `var_explained`: proportion of variance explained by PC1.
#'   }
#' @importFrom stats sd cov
#' @export
naive_pca <- function(S_hat) {
  N <- nrow(S_hat)
  K <- ncol(S_hat)

  col_means <- colMeans(S_hat)
  col_sds   <- apply(S_hat, 2, sd)

  # Standardize
  S_tilde <- sweep(sweep(S_hat, 2, col_means, "-"), 2, col_sds, "/")

  # Sample correlation matrix
  R_obs <- (t(S_tilde) %*% S_tilde) / (N - 1)

  # Leading eigenpair
  ev      <- eigen(R_obs, symmetric = TRUE)
  lambda1 <- ev$values[1]
  ell_hat <- ev$vectors[, 1]

  # Sign convention: choose direction where sum of loadings is positive
  # (in practice all loadings are positive when sub-indices share a common factor)
  if (sum(ell_hat) < 0) ell_hat <- -ell_hat

  # Regression scoring coefficients: w = R_obs^{-1} * ell
  w_hat <- as.vector(solve(R_obs) %*% ell_hat)

  # Composite scores
  f_hat <- as.vector(S_tilde %*% w_hat)

  # Variance explained
  var_explained <- lambda1 / sum(ev$values)

  list(
    col_means    = col_means,
    col_sds      = col_sds,
    S_tilde      = S_tilde,
    R_obs        = R_obs,
    lambda1      = lambda1,
    ell_hat      = ell_hat,
    w_hat        = w_hat,
    f_hat        = f_hat,
    var_explained = var_explained
  )
}
